<html><head><title>Parsed Arguments</title></head><body><h1>Flags available for pre-processing (ToBridge)</h1><table><tr><th>Flag</th><th>Default</th><th>Help</th></tr><tr><td>--cr_mkfastq</td><td>False</td><td>Run cellranger make fastq; make sure SampleSheet.csv is in your BCL folders</td></tr><tr><td>--bcl_convert</td><td>False</td><td>Run BCL convert; make sure SampleSheet.csv is in your BCL folders</td></tr><tr><td>--bcl_convert_sheet_conv</td><td>False</td><td>Convert sample sheet to use BCL convert; use this flag if you have a 10x-style sample sheet</td></tr><tr><td>--fastqc</td><td>False</td><td>Run FASTQC</td></tr><tr><td>--star_solo</td><td>False</td><td>Run STARsolo</td></tr><tr><td>--star_solo_chem</td><td>None</td><td>What is the sample chemistry? If you run STARsolo, this must be provided. Example: SC3Pv3</td></tr><tr><td>--star_solo_features</td><td>GeneFull_Ex50pAS</td><td>Quantification of different features; currently GeneFull_Ex50pAS (closest to CellRanger) and SJ (splicing analysis) have been tested. You can use them together as 'GeneFull_Ex50pAS SJ'</td></tr><tr><td>--star_solo_genome_generate</td><td>False</td><td>Generate a reference genome for STARsolo</td></tr><tr><td>--star_solo_bam</td><td>True</td><td>Generate sorted BAM files instead of SAMs</td></tr><tr><td>--cr_count</td><td>False</td><td>Run cellranger count</td></tr><tr><td>--cr_genome_generate</td><td>None</td><td>Generate a new cellranger reference genome. Please provide a desired name</td></tr><tr><td>--cr_count_chemistry</td><td>None</td><td>Specify chemistry for cellranger if cannot be detected automatically (rare)</td></tr><tr><td>--cr_count_forcecells</td><td>None</td><td>Force a certain number of cells for cellranger count</td></tr><tr><td>--cr_count_bam</td><td>False</td><td>Produce BAM files upon alignment. Note that the default is to not produce BAMs</td></tr><tr><td>--cr_count_feature</td><td>False</td><td>Feature bacrode assay using cellranger</td></tr><tr><td>--alignment_qc</td><td>False</td><td>Generates a QC report based on alignment results</td></tr><tr><td>--alignment_qc_genebody</td><td>False</td><td>Generates a QC report based on alignment results, which includes genebody coverage (slow)</td></tr><tr><td>--multi_qc</td><td>True</td><td>Generates an HTML report based on all the QC</td></tr><tr><td>--samp</td><td>All</td><td>If you only want cellranger run on specific samples, use this flag, e.g. --samp S1,S5</td></tr><tr><td>--nthreads</td><td>cpu_count-1</td><td>Number of threads; defaults to all minus 1</td></tr></table></body></html>